staging/src/k8s.io/client-go/rest/request.go

```diff
@@ -786,22 +787,47 @@ func (r *Request) newStreamWatcher(resp *http.Response) (watch.Interface, error)
 	), nil
 }
 
-// updateURLMetrics is a convenience function for pushing metrics.
-// It also handles corner cases for incomplete/invalid request data.
-func updateURLMetrics(ctx context.Context, req *Request, resp *http.Response, err error) {
-	url := "none"
+// updateRequestResultMetric increments the RequestResult metric counter,
+// it should be called with the (response, err) tuple from the final
+// reply from the server.
+func updateRequestResultMetric(ctx context.Context, req *Request, resp *http.Response, err error) {
+	code, host := sanitize(req, resp, err)
+	metrics.RequestResult.Increment(ctx, code, req.verb, host)
+}
+
+// updateRequestRetryMetric increments the RequestRetry metric counter,
+// it should be called with the (response, err) tuple for each retry
+// except for the final attempt.
+func updateRequestRetryMetric(ctx context.Context, req *Request, resp *http.Response, err error) {
+	code, host := sanitize(req, resp, err)
+	metrics.RequestRetry.IncrementRetry(ctx, code, req.verb, host)
+}
+
+func updateRequestRetryOrResultMetric(ctx context.Context, done bool, req *Request, resp *http.Response, err error) {
+	if !done {
+		// this is not the final attempt, there is at least one more
+		// attempt, so count this attempt as a retry.
+		updateRequestRetryMetric(ctx, req, resp, err)
+	}
```

Comment:
If we only record the final attempt as a retry, all the previous retries will be not be accounted for

Reply:
this is the semantics of the result and retry metric, i added it as a comment
```
	// the client finishes a single request after N attempts (1..N)
	//  - all attempts (1..N) are counted to the rest_client_requests_total
	//    metric (current behavior).
	//  - every attempt after the first (2..N) are counted to the
	//    rest_client_request_retries_total metric.
```